source/functional-concepts.md

```diff
@@ -0,0 +1,150 @@
+# Functional Programming in Swift: A Primer
+
+Swift is a hybrid language that allows both objective-oriented and functional code to co-exist.
+
+### First-class functions
+
+First-class functions are *values* that can either take other functions as arguments or return them as results. Also refered to as *higher order functions*.
+
+#### Returning Functions
+
+~~~swift
+typealias Prefixer = String -> String
+
+func prefixerFactory(prefix: String) -> Prefixer {
+// func prefix(prefix: String) -> (String -> String)
+
+  func aPrefixer(word: String) -> String {
+    return prefix + word
+  }
+
+  return aPrefixer
+}
+
+let repeat = prefixerFactory("re-")
+repeat("do")
+~~~
+
+*Note*: You can also return an unnamed function, i.e. a closure.
+
+~~~swift
+func otherPrefixerfactory(prefix: String) -> Prefixer {
+  return { word in
+    return prefix + word
+  }
+}
+~~~
+
+#### Taking Functions as Arguments
+
+~~~swift
+func uppercasePrefixer(word: String, prefixer: Prefixer) -> String {
+  return prefixer(word).uppercaseString
+}
+
+uppercasePrefixer("do", repeat)
+~~~
+
+### Functional Purity
+
+Functions should have no side effects. They do not alter the state around them:
+
+#### Immutability
+
+Functions operate on copies of existing values instead of mutating them.
+
+~~~swift
+struct Employee {
+  let name: String
+  var age: Int
```

Comment:
@tonyd256 I want it to be a variable to drive the point. Making it a constant will make `mutating` look off but for the wrong reasons.